src/main/java/com/npe/pet/uncaughtHeroes/controller/HeroController.java

```diff
@@ -0,0 +1,56 @@
+package com.npe.pet.uncaughtHeroes.controller;
+
+import com.npe.pet.uncaughtHeroes.entity.Hero;
+import com.npe.pet.uncaughtHeroes.service.HeroService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/api/heroes")
+public class HeroController {
+
+    private final HeroService heroService;
+
+    @Autowired
+    public HeroController(HeroService heroService) {
+        this.heroService = heroService;
+    }
+
+    @PostMapping
+    public ResponseEntity<Hero> saveHero(@RequestBody Hero hero) {
+        Optional<Hero> savedHero = heroService.save(hero);
+        return savedHero.map(value -> new ResponseEntity<>(value, HttpStatus.CREATED))
```

Comment:
Here, I believe that an if-else structure would be easier to read and understand. 

This also works fine, but in getHeroById, you also used an if-else structure for pretty much the same thing, and I believe it would be better if we would use the same structure across the whole controller. So I would suggest that either change saveHero's implementation to an if-else, or change getHeroByID to use a similar stream structure.

Reply:
First we should agree on where we are going to handle Optionals. Right now for the save operation the optional is "unwrapped" in the controller, for findById it is handled in the service. I could argue for both, for example if we handled in the service methods, we wouldn't need to worry about Optionals in the unit tests. If we handle it in the controller, then our service methods would never return null.

What do you think?

Reply:
As a third opinion I would say that a controller should not mess with unwrapping the optional. The service should return the actual object or throw an exception. 

Or an other alternative is to unwrap a the hero object like this.

Hero hero = savedHero.get().else(null); 
or .elseThrow(new AnyExceptionYouLike()) and handle the exception in a ControllerAdvice
return ...